File: leetcode-go/Q079_Word_Search/Q079_Word_Search.go
package main //Q079_Word_Search

import "fmt"

func exist(board [][]byte, word string) bool {
	if len(board) == 0 {
		return len(word) == 0
	}

	m, n := len(board), len(board[0])
	visit := make([][]bool, m)
	for i := 0; i < m; i++ {
		visit[i] = make([]bool, n)
	}

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			visit[i][j] = true
			if dfs(board, visit, i, j, []byte(word)) {
				return true
			}
			visit[i][j] = false
		}
	}

	return false
}

func dfs(board [][]byte, visit [][]bool, x, y int, word []byte) bool {
	if len(word) == 1 {
		return board[x][y] == word[0]
	}
	if board[x][y] != word[0] {
		return false
	}
	dirs := [][]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}
	for i := 0; i < 4; i++ {
		xx := x + dirs[i][0]
		yy := y + dirs[i][1]
		if xx < 0 || xx > len(board)-1 || yy < 0 || yy > len(board[0])-1 {
			continue
		}
		if visit[xx][yy] {
			continue
		}
		visit[xx][yy] = true
		if dfs(board, visit, xx, yy, word[1:]) {
			return true
		}
		visit[xx][yy] = false
	}
	return false
}

func main() {
	board := [][]byte{
		[]byte("ABCE"),
		[]byte("SFCS"),
		[]byte("ADEE"),
	}
	fmt.Println(exist(board, "ABCCED"))
}

File: src/facebook/backtracking/Q079_Word_Search.java
package facebook.backtracking;

public class Q079_Word_Search {

    int[][] dirs = new int[][]{{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

    public boolean exist(char[][] board, String word) {
        if (board == null || board.length == 0 || board[0] == null || board[0].length == 0) return false;
        int m = board.length, n = board[0].length;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (dfs(board, i, j, word, 0, new boolean[m][n])) return true;
            }
        }
        return false;
    }

    private boolean dfs(char[][] board, int i, int j, String word, int pos, boolean[][] visited) {
        int m = board.length, n = board[0].length;
        if (pos == word.length()) return true;
        if (i < 0 || i >= m || j < 0 || j >= n || visited[i][j]) return false;
        if (word.charAt(pos) != board[i][j]) return false;

        visited[i][j] = true;
        for (int[] dir : dirs) {
            int ii = i + dir[0], jj = j + dir[1];
            if (dfs(board, ii, jj, word, pos + 1, visited)) return true;
        }
        visited[i][j] = false;
        return false;
    }

}
